In [16]:
# import the necessary packages
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import accuracy_score,precision_score,confusion_matrix,classification_report, precision_recall_curve
from sklearn.svm import SVC
import numpy as np

print("Loading libraries    ...")

Loading libraries    ...


In [17]:
# Load the data from the file
# File is in the ../Data directory
# Fine name is "cleaned_students.csv"

cleanedStudentsGradesDf = pd.read_csv("../data/cleaned_students.csv")
cleanedStudentsGradesDf.head()

Unnamed: 0 grading_scale  region_East Anglian Region  \
0           0            C-                           1   
1           1            A-                           0   
2           2            A-                           0   
3           3             B                           0   
4           4            A+                           0   

   region_East Midlands Region  region_Ireland  region_London Region  \
0                            0               0                     0   
1                            0               0                     0   
2                            0               0                     0   
3                            0               0                     0   
4                            0               0                     0   

   region_North Region  region_North Western Region  region_Scotland  \
0                    0                            0                0   
1                    0                            0                1   
2                    0                            0                0   
3                    0                            0                0   
4                    0                            0                0   

   region_South East Region  ...  module_presentation_length_262  \
0                         0  ...                               0   
1                         0  ...                               0   
2                         1  ...                               0   
3                         0  ...                               0   
4                         0  ...                               0   

   module_presentation_length_268  module_presentation_length_269  \
0                               1                               0   
1                               1                               0   
2                               1                               0   
3                               1                               0   
4                               1                               0   

   date_average_binned_0-30  date_average_binned_30-60  \
0                         0                          0   
1                         0                          0   
2                         0                          1   
3                         0                          0   
4                         0                          0   

   date_average_binned_60-90  date_average_binned_120-150  \
0                          0                            0   
1                          0                            0   
2                          0                            0   
3                          0                            0   
4                          0                            0   

   date_average_binned_150-180  date_average_binned_180-210  \
0                            0                            0   
1                            0                            0   
2                            0                            0   
3                            0                            0   
4                            0                            0   

   date_average_binned_210-240  
0                            0  
1                            0  
2                            0  
3                            0  
4                            0  

[5 rows x 54 columns]

In [18]:
# Divide the X and y coordinates into two dataframes for training. gradingScale can be the labels dataframe
# and the rest can be the features dataframe
gradingScaleLabelsDf = cleanedStudentsGradesDf['grading_scale']
gradingScaleLabelsDf.head()

featuresDf = cleanedStudentsGradesDf.drop(['grading_scale', 'Unnamed: 0'], axis=1)
featuresDf.head()


region_East Anglian Region  region_East Midlands Region  region_Ireland  \
0                           1                            0               0   
1                           0                            0               0   
2                           0                            0               0   
3                           0                            0               0   
4                           0                            0               0   

   region_London Region  region_North Region  region_North Western Region  \
0                     0                    0                            0   
1                     0                    0                            0   
2                     0                    0                            0   
3                     0                    0                            0   
4                     0                    0                            0   

   region_Scotland  region_South East Region  region_South Region  \
0                0                         0                    0   
1                1                         0                    0   
2                0                         1                    0   
3                0                         0                    0   
4                0                         0                    0   

   region_South West Region  ...  module_presentation_length_262  \
0                         0  ...                               0   
1                         0  ...                               0   
2                         0  ...                               0   
3                         0  ...                               0   
4                         0  ...                               0   

   module_presentation_length_268  module_presentation_length_269  \
0                               1                               0   
1                               1                               0   
2                               1                               0   
3                               1                               0   
4                               1                               0   

   date_average_binned_0-30  date_average_binned_30-60  \
0                         0                          0   
1                         0                          0   
2                         0                          1   
3                         0                          0   
4                         0                          0   

   date_average_binned_60-90  date_average_binned_120-150  \
0                          0                            0   
1                          0                            0   
2                          0                            0   
3                          0                            0   
4                          0                            0   

   date_average_binned_150-180  date_average_binned_180-210  \
0                            0                            0   
1                            0                            0   
2                            0                            0   
3                            0                            0   
4                            0                            0   

   date_average_binned_210-240  
0                            0  
1                            0  
2                            0  
3                            0  
4                            0  

[5 rows x 52 columns]

In [21]:
# Check the shape of the features and labels dataframes
print("Features shape:", featuresDf.shape)
print("Labels shape:", gradingScaleLabelsDf.shape)

# Ensure that the features and labels dataframes have the same number of rows
if featuresDf.shape[0] != gradingScaleLabelsDf.shape[0]:
    raise ValueError("Number of rows in features and labels dataframes are not equal.")

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(featuresDf, gradingScaleLabelsDf, test_size=0.3)


Features shape: (32560, 52)
Labels shape: (32560,)


In [32]:
# Create a logistic regression with multi_class set to ovr and the solver set to liblinear

# logisticRegr = LogisticRegression(multi_class='ovr', solver='liblinear')
# logisticRegr = LogisticRegression(multi_class='ovr', solver='lbfgs')
# logisticRegr = LogisticRegression(multi_class='ovr', solver='newton-cg')
# logisticRegr = LogisticRegression(multi_class='ovr', solver='sag')
# logisticRegr = LogisticRegression(multi_class='ovr', solver='saga')
logisticRegr = LogisticRegression(multi_class='ovr', solver='newton-cholesky')



model = logisticRegr.fit(X_train, np.ravel(y_train))
accuracy = model.score(X_test, y_test)

# Print the accuracy of the model in % format with 2 decimal places
print("Accuracy is {:.2f}%".format(accuracy * 100))


Accuracy is 17.24%


In [27]:
# Get more detail by printing a classification report
y_pred = model.predict(X_test)
print(classification_report(y_test,y_pred))


              precision    recall  f1-score   support

           A       0.20      0.23      0.22       854
          A+       0.14      0.12      0.13       892
          A-       0.16      0.30      0.21       898
           B       0.09      0.01      0.02       953
          B+       0.16      0.11      0.13       860
          B-       0.16      0.06      0.08       840
           C       0.12      0.05      0.07       905
          C+       0.17      0.10      0.13       898
          C-       0.21      0.43      0.28       874
           D       0.14      0.25      0.18       878
           F       0.21      0.26      0.23       916

    accuracy                           0.17      9768
   macro avg       0.16      0.17      0.15      9768
weighted avg       0.16      0.17      0.15      9768



Linear Regression Results

Vary solver:

| Solver | Accuracy [%] |
|-----------------|-----------------|
| liblinear | 17.26 |
| lbfgs | 17.24 |
| newton-cg | 17.24 |
| sag | 17.24 |
| saga | 17.24 |
| newton-cholesky | 17.24 |



